File: frontend/src/lib/certificate-layout.ts
export type CertificateFieldAlign = "left" | "center" | "right"

export interface CertificateFieldLayout {
  x_ratio: number
  y_ratio: number
  font_ratio: number
  align: CertificateFieldAlign
}

export interface CertificateLayout {
  name: CertificateFieldLayout
  id: CertificateFieldLayout
  text_color: string
}

export const SAMPLE_CERTIFICATE_NAME = "John Doe"
export const SAMPLE_CERTIFICATE_ID = "2024001234"

export function defaultCertificateLayout(): CertificateLayout {
  return {
    name: { x_ratio: 0.5, y_ratio: 0.42, font_ratio: 0.038, align: "center" },
    id: { x_ratio: 0.5, y_ratio: 0.48, font_ratio: 0.024, align: "center" },
    text_color: "#1a1a1a",
  }
}

function clampRatio(value: number, low = 0.02, high = 0.98): number {
  return Math.max(low, Math.min(high, value))
}

function normalizeField(raw: unknown, defaults: CertificateFieldLayout): CertificateFieldLayout {
  const field = { ...defaults }
  if (!raw || typeof raw !== "object") return field
  const obj = raw as Record<string, unknown>
  if (typeof obj.x_ratio === "number") field.x_ratio = clampRatio(obj.x_ratio)
  if (typeof obj.y_ratio === "number") field.y_ratio = clampRatio(obj.y_ratio)
  if (typeof obj.font_ratio === "number") field.font_ratio = Math.max(0.01, Math.min(0.12, obj.font_ratio))
  if (obj.align === "left" || obj.align === "center" || obj.align === "right") field.align = obj.align
  return field
}

/** Merge API / legacy flat keys into the canonical nested layout shape. */
export function normalizeCertificateLayout(raw?: Record<string, unknown> | null): CertificateLayout {
  const base = defaultCertificateLayout()
  if (!raw) return base

  const result: CertificateLayout = {
    name: normalizeField(raw.name, base.name),
    id: normalizeField(raw.id, base.id),
    text_color: typeof raw.text_color === "string" && raw.text_color ? raw.text_color : base.text_color,
  }

  if (typeof raw.name_y_ratio === "number") result.name.y_ratio = clampRatio(raw.name_y_ratio)
  if (typeof raw.name_font_ratio === "number") result.name.font_ratio = Math.max(0.01, Math.min(0.12, raw.name_font_ratio))
  if (typeof raw.name_x_ratio === "number") result.name.x_ratio = clampRatio(raw.name_x_ratio)

  if (typeof raw.id_y_ratio === "number") result.id.y_ratio = clampRatio(raw.id_y_ratio)
  if (typeof raw.id_font_ratio === "number") result.id.font_ratio = Math.max(0.01, Math.min(0.12, raw.id_font_ratio))
  if (typeof raw.id_x_ratio === "number") result.id.x_ratio = clampRatio(raw.id_x_ratio)

  return result
}

export function fieldTransform(align: CertificateFieldAlign): string {
  if (align === "left") return "translate(0, -50%)"
  if (align === "right") return "translate(-100%, -50%)"
  return "translate(-50%, -50%)"
}

export function fieldFontSize(fontRatio: number, containerHeight: number): string {
  const px = Math.max(10, Math.round(containerHeight * fontRatio))
  return `${px}px`
}

File: backend/main/services/certificate_layout.py
"""Normalized certificate text layout (template overlay positions)."""

from __future__ import annotations

import copy


def default_certificate_layout() -> dict:
    """Relative coordinates (0–1) for name and ID on the template image."""
    return {
        "name": {
            "x_ratio": 0.5,
            "y_ratio": 0.42,
            "font_ratio": 0.038,
            "align": "center",
        },
        "id": {
            "x_ratio": 0.5,
            "y_ratio": 0.48,
            "font_ratio": 0.024,
            "align": "center",
        },
        "text_color": "#1a1a1a",
    }


def _clamp_ratio(value: float, *, low: float = 0.02, high: float = 0.98) -> float:
    return max(low, min(high, value))


def _normalize_field(raw: dict | None, defaults: dict) -> dict:
    field = copy.deepcopy(defaults)
    if not isinstance(raw, dict):
        return field
    if "x_ratio" in raw:
        field["x_ratio"] = _clamp_ratio(float(raw["x_ratio"]))
    if "y_ratio" in raw:
        field["y_ratio"] = _clamp_ratio(float(raw["y_ratio"]))
    if "font_ratio" in raw:
        field["font_ratio"] = max(0.01, min(0.12, float(raw["font_ratio"])))
    align = raw.get("align")
    if align in ("left", "center", "right"):
        field["align"] = align
    return field


def normalize_certificate_layout(raw: dict | None) -> dict:
    """
    Merge stored layout with defaults. Supports legacy flat keys
    (name_y_ratio, id_font_ratio, …) and the nested { name, id } shape.
    """
    base = default_certificate_layout()
    if not raw:
        return base

    result = copy.deepcopy(base)
    result["name"] = _normalize_field(raw.get("name"), base["name"])
    result["id"] = _normalize_field(raw.get("id"), base["id"])

    if "text_color" in raw and raw["text_color"]:
        result["text_color"] = str(raw["text_color"])

    if "name_y_ratio" in raw:
        result["name"]["y_ratio"] = _clamp_ratio(float(raw["name_y_ratio"]))
    if "name_font_ratio" in raw:
        result["name"]["font_ratio"] = max(0.01, min(0.12, float(raw["name_font_ratio"])))
    if "name_x_ratio" in raw:
        result["name"]["x_ratio"] = _clamp_ratio(float(raw["name_x_ratio"]))

    if "id_y_ratio" in raw:
        result["id"]["y_ratio"] = _clamp_ratio(float(raw["id_y_ratio"]))
    if "id_font_ratio" in raw:
        result["id"]["font_ratio"] = max(0.01, min(0.12, float(raw["id_font_ratio"])))
    if "id_x_ratio" in raw:
        result["id"]["x_ratio"] = _clamp_ratio(float(raw["id_x_ratio"]))

    return result
